scripts/kb.py

```diff
@@ -0,0 +1,633 @@
+        self,
+        vocab: Vocab,
+        entity_vector_length: int,
+        db_path: Path,
+        annoy_path: Path,
+        language: str,
+        n_trees: int = 25,
+        top_k_aliases: int = 5,
+        top_k_entities_alias: int = 20,
+        top_k_entities_fts: int = 5,
+        threshold_alias: int = 100,
+    ):
+        """Initializes from existing SQLite database generated by `wikid`.
+        Loads Annoy index file (as mmap) into memory, if file exists at specified path.
+        vocab (Vocab): Pipeline vocabulary.
+        entity_vector_length (int): Length of entity vectors.
+        db_path (Path): Path to SQLite database.
+        annoy_path (Path): Path to Annoy index file.
+        language (str): Language.
+        n_trees (int): Number of trees in Annoy index. Precision in NN queries correspond with number of trees. Ignored
+            if Annoy index file already exists.
+        top_k_aliases (int): Top k aliases matches to consider. An alias may be associated with more than one entity, so
+            this parameter does _not_ necessarily correspond to the the maximum number of identified candidates. For
+            that use top_k_alias_entity.
+        top_k_alias_entity (int): Top k entities to consider in list of alias matches. Equals maximum number of
+            candidate entities found via alias search.
+        top_k_entities_fts (int): Top k of full-text search matches to consider. Equals maximum number of candidate entities
+            found via full-text search.
+        threshold_alias (int): Threshold for alias distance as calculated by spellfix1.
+        """
+        super().__init__(vocab, entity_vector_length)
+
+        self._paths = {"db": db_path, "annoy": annoy_path}
+        self._language = language
+        self._annoy: Optional[annoy.AnnoyIndex] = None
+        self._n_trees = n_trees
+        self._db_conn = establish_db_connection(language, self._paths["db"])
+        self._hashes: Dict[str, Optional[str]] = {}
+        self._top_k_aliases = top_k_aliases
+        self._top_k_entities_alias = top_k_entities_alias
+        self._top_k_entities_fts = top_k_entities_fts
+        self._threshold_alias = threshold_alias
+
+        if os.path.exists(self._paths["annoy"]):
+            self._init_annoy_from_file()
+
+    def build_embeddings_index(
+        self, nlp: Language, n_jobs: int = -1, batch_size: int = 2**14
+    ) -> None:
+        """Constructs index for embeddings with Annoy and stores them in an index file.
+        nlp (Language): Pipeline with tok2vec for inferring embeddings.
+        n_jobs (int): Number of jobs to use for inferring entity embeddings and building the index.
+        batch_size (int): Number of entities to request at once.
+        """
+        logger = logging.getLogger(__name__)
+
+        # Initialize ANN index.
+        self._annoy = annoy.AnnoyIndex(self.entity_vector_length, "angular")
+        self._annoy.on_disk_build(str(self._paths["annoy"]))
```

Comment:
Should this warn/exit when `self._annoy` is already defined/available?

Reply:
Updated in https://github.com/explosion/wikid/pull/4/commits/33d6db297cb5d8bf9ee85023c604b8f9e236cd3e and https://github.com/explosion/wikid/pull/4/commits/eb80b708eca1ea97bb10bb81626aa08a850f72d0. Fails now if index is available.